src/Host/Client/Impl/BrokerServices/WebServer.cs

```diff
@@ -0,0 +1,155 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Net;
+using System.Net.Http;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Common.Core;
+using Newtonsoft.Json;
+
+namespace Microsoft.R.Host.Client.BrokerServices {
+    public class WebServer {
```

Comment:
We have discussed this approach at length. The arrangement with the local server is the only one that works reliably. Doing it on broker side actually ends up being more complicated, because then you have to rewrite URLs in the content (so that `/foo` becomes `/remoteuri/8080/foo` etc - don't forget that you'd need to encode the port also, since there may be multiple servers run by multiple users on the broker). Doing this reliably for HTML and JS and whatever else might be on the wire is pretty much impossible in general.

Host is definitely the wrong place to do it, because it would require every new host instance to open a port of its own for direct remote connections. We don't want to tell server admins that there may potentially be several dozen random ports open on their box when multiple users are running their stuff.